File: 101/ruby/linked_list.rb
# frozen_string_literal: true

# Linked list implementation.
class LinkedList
  attr_reader :head

  # Add an object to the tail of the list.
  def add(data)
    node = LinkedListNode.new(data)

    @head ||= node

    @tail.next_node = node unless @tail.nil?
    @tail = node
  end

  # Add to the front of the list.
  def add_first(data)
    node = LinkedListNode.new data

    node.next_node = @head
    @head = node
  end

  # Get the element at a specific index, returning nil if there is no item at that index.
  def get(index)
    get_node(index)&.data
  end

  # Get the node holding the element at a specific index, returning nil if there is no item at that index.
  def get_node(index)
    find_cursor(index)
  end

  # Delete from the list.
  def delete(index)
    cursor = find_cursor(index - 1)
    return if cursor.nil? || cursor.next_node.nil?

    cursor.next_node = cursor.next_node&.next_node
    @tail = cursor.next_node if cursor == @tail
  end

  # Find the first occurrence of a term within the linked list.
  def find(term)
    node = @head

    until node.nil?
      return true if node.data == term

      node = node.next_node
    end

    false
  end

  # Find all occurrences of a term in the linked list.
  def collect
    arr = []
    node = @head

    until node.nil?
      arr << node.data if yield(node.data)
      node = node.next_node
    end

    arr
  end

  # Sort the linked list.
  def sort # rubocop:disable Metrics/MethodLength
    list = @head

    until list.nil?
      pass = list.next_node

      until pass.nil?
        if list.data <=> pass.data
          tmp = list.data
          list.data = pass.data
          pass.data = tmp
        end

        pass = pass.next_node
      end

      list = list.next_node
    end
  end

  # Reverse the linked list.
  def reverse
    current_node = @head
    prev_node = nil

    until current_node.nil?
      next_node = current_node.next_node
      current_node.next_node = prev_node
      prev_node = current_node
      current_node = next_node
    end

    @tail = @head
    @head = prev_node
  end

  # Traverse the linked list.
  def traverse
    cursor = @head

    while cursor
      yield cursor.data
      cursor = cursor.next_node
    end
  end

  # Check if there are cycles, or infinite loops, in this linked list.
  def cycles?
    check_for_cycles(@head, @head)
  end

  private

  # Position the cursor pointing to a desired index.
  def find_cursor(index)
    cursor = @head
    index.times { cursor = cursor&.next_node }
    cursor
  end

  # Check if there are cycles, or infinite loops, in this linked list.
  def check_for_cycles(slow, fast)
    collision = false
    while !slow.nil? && !fast.nil? && !fast.next_node.nil?
      slow = slow.next_node
      fast = fast.next_node.next_node
      if slow == fast
        collision = true
        break
      end
    end
    collision
  end
end

# Linked list node implementation.
class LinkedListNode
  attr_accessor :data, :next_node

  def initialize(data)
    @data = data
  end
end

File: 101/ts/src/linkedList.ts
class Node<T> {
  data: T;
  next?: Node<T>;

  /**
   * A node in a linked list.
   *
   * @param obj the object the node in the list represents.
   * @constructor
   */
  constructor(data: T, next?: Node<T>) {
    this.data = data;
    this.next = next;
  }
}

/**
 * A linked list.
 */
export class LinkedList<T> {
  head?: Node<T>;
  tail?: Node<T>;

  /**
   * Add an object to the tail of the list.
   *
   * @param {T} value the object to add.
   */
  add = (value: T): void => {
    const node = new Node(value);

    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else if (this.tail) {
      this.tail.next = node;
      this.tail = node;
    }
  };

  /**
   * Add an object to the list at a specified index.
   *
   * @param {number} idx the index to insert the new node.
   * @param {T} value the object to add.
   */
  addAtIndex = (idx: number, value: T): void => {
    const insertAt = this.getNode(idx - 1);

    if (insertAt) {
      const newNode = new Node(value, insertAt.next);
      insertAt.next = newNode;

      if (insertAt === this.tail) {
        this.tail = newNode;
      }
    }
  };

  /**
   * Add an object to the front of the list.
   *
   * @param {T} value the object to add.
   */
  addFirst = (value: T): void => {
    this.head = new Node(value, this.head);
  };

  /**
   * Get an item from the linked list.
   *
   * @param index the index to get.
   */
  get = (idx: number): T | undefined => this.getNode(idx)?.data;

  /**
   * Find the cursor pointing at a node in the linked list.
   *
   * @param {number} idx the index of the node to find the cursor for.
   * @returns {*} the cursor, if the index is valid.
   */
  getNode = (idx: number): Node<T> | undefined => {
    let i = 0;
    let curr = this.head;

    while (i <= idx && curr) {
      if (i === idx) return curr;
      i += 1;
      curr = curr.next;
    }

    return undefined;
  };

  /**
   * Check if the linked list has cycles, or infinite traversal loops.
   *
   * @returns {Boolean}
   */
  hasCycles = (): boolean => {
    let slow = this.head;
    let fast = this.head;
    let collision = false;

    while (slow && fast && fast?.next) {
      slow = slow?.next;
      fast = fast?.next?.next;
      if (slow === fast) {
        collision = true;
        break;
      }
    }

    return collision;
  };

  /**
   * Remove an item from the linked list.
   *
   * @param idx the index to remove.
   */
  remove = (idx: number): boolean => {
    const prev = this.getNode(idx - 1);

    if (prev) {
      prev.next = prev.next?.next;
      return true;
    }

    return false;
  };

  /**
   * Reverse the linked list.
   */
  reverse = (): void => {
    let currentNode = this.head;
    let previousNode, nextNode;

    while (currentNode) {
      nextNode = currentNode.next;
      currentNode.next = previousNode;
      previousNode = currentNode;
      currentNode = nextNode;
    }

    this.tail = this.head;
    this.head = previousNode;
  };

  /**
   * Sort the linked list.
   */
  sort = (): void => {
    let list = this.head;

    while (list) {
      let pass = list.next;

      while (pass) {
        if (list.data < pass.data) {
          const tmp = list.data;
          list.data = pass.data;
          pass.data = tmp;
        }

        pass = pass.next;
      }

      list = list.next;
    }
  };

  /**
   * Traverse the linked list.
   *
   * @param {function} callback called with each item in the linked list.
   */
  traverse = (callback: (data: T) => void): void => {
    let curr = this.head;
    while (curr) {
      callback(curr.data);
      curr = curr.next;
    }
  };
}
